Make TimedReloadStrategy error-prone and reusable

detouched opened the pull request:
Current `TimedReloadStrategy` has two flaws:
* the underlying `Timer` terminates if any reload attempt fails with an exception. This makes this reload strategy suck when dealing with I/O, for instance, reloading from a remote URL.
* if `register(ConfigProvider)` is called multiple times, only the last `ConfigProvider` can effectively call `deregister()`, while all previously registered timers will be lost. This makes the reload strategy not reusable, while its API suggests that it might be reused.

This change makes `TimedReloadStrategy` use a scheduled executor service with error-prone `reload()` call to prevent subsequent reload attempts from auto-cancelling, and use a map of `ConfigProvider`s to register and unregister them properly.

detouched committed:
Make TimedReloadStrategy error-prone and reusable

- TimedReloadStrategy does not abort in case the ConfigProvider
  throws an exception on reload()
- TimedReloadStrategy can serve multiple ConfigProviders

detouched commented on cfg4k-core/src/main/kotlin/com/jdiazcano/cfg4k/reloadstrategies/TimedReloadStrategy.kt:
```diff
@@ -17,25 +17,47 @@
-    private lateinit var reloadTimer: Timer
+    private val executor: ScheduledExecutorService by lazy {
+        Executors.newSingleThreadScheduledExecutor { runnable ->
+            Thread(runnable, "TimeReloadStrategy").also { it.isDaemon = true }
```
Setting executor's thread as a `daemon` so that it doesn't prevent JVM shut down.

detouched commented on cfg4k-core/src/main/kotlin/com/jdiazcano/cfg4k/reloadstrategies/TimedReloadStrategy.kt:
```diff
@@ -17,25 +17,47 @@
+            val reload = {
+                try {
+                    cp.reload()
+                } catch (ignored: Exception) {
```
Nothing good in swallowing the exception here, but since the rest of the `cfg4k` doesn't log errors, I decided to leave this as is.
I understand that additional dependencies are undesirable, but I can make use of JUL as a workaround ¯\\\_(ツ)\_/¯ Let me know what you think about this.

jdiazcano replied:
The config provider must handle the exceptions so I would say there is no need to have this try catch here. 

As for the exceptiosn there is an `addReloadErrorListener()` which these listeners will be called when an exception is thrown when loading/reloading.

* [addReloadErrorListener interface](https://github.com/jdiazcano/cfg4k/blob/master/cfg4k-core/src/main/kotlin/com/jdiazcano/cfg4k/providers/ConfigProvider.kt#L85)
* [DefaultConfigProvider notify errors](https://github.com/jdiazcano/cfg4k/blob/master/cfg4k-core/src/main/kotlin/com/jdiazcano/cfg4k/providers/DefaultConfigProvider.kt#L129)

detouched replied:
Hm, so this is a recent work as I can see, it has not been included into any release so far.
Sorry, I found the issues I tried to address here while working on a pet project with cfg4k added as a dependency, so I was mostly looking into source of version 0.8 and then ported my change and added tests in the actual cfg4k repo.

Cool, so if `ConfigProvider` now has a contract of catching exceptions on reload and notifying possible listeners, then my change is indeed redundant.

jdiazcano replied:
There's no release yet but I could publish `0.8.1` once this is merged.

detouched commented:
The build is failing due to some openssl pre-configuration failure, not sure how to fix that but I guess it's unrelated to the proposed change.

jdiazcano commented:
@detouched Hello!

Thanks for taking your time of doing this!.

The build is failing because the test that executes the Git checkout via SSH can't be decrypted with the pull requests that aren't coming from me. This is something I really need to check. But it is totally unrelated, it is due on how travis-ci works with encrypted files.

I'm leaving some comments before pulling it in but I like the shape of it.

jdiazcano commented on cfg4k-core/src/test/kotlin/com/jdiazcano/cfg4k/TimedReloadStrategyTest.kt:
```diff
@@ -78,8 +81,47 @@ nested.a=reloaded nestedb
+            val lastSeenReloadCounter2 = reloadCounter2
+            Thread.sleep(100)
+            reloadStrategy.deregister(provider2)
+            val finishedReloadCounter2 = reloadCounter2
```
This variable doesn't seem to be used. Is this a miss or can be removed?

detouched replied:
Yes, a leftover. Interesting, it didn't appear as not used but it's obviously not used 🤔

jdiazcano commented on cfg4k-core/src/test/kotlin/com/jdiazcano/cfg4k/TimedReloadStrategyTest.kt:
```diff
@@ -78,8 +81,47 @@ nested.a=reloaded nestedb
         }
-    }
 
+        it("error on reload doesn't prevent further reload attempts") {
```
I like this test!

detouched replied:
Well, now that the contract of `ConfigProvider` is to not to throw any errors, and the try-catch above is a surplus, this test is becomes wrong. So for this PR I'm going to remove it.
However, it might be reused to test catching of errors in the `DefaultConfigProvider.reload()` you mentioned above.

jdiazcano replied:
I think it is good to leave it there, it proves that we are catching the exception when something happens. I think this was a miss on my side when added the reload error listener.

jdiazcano reviewed (changes requested):
I would say to just remove (or discuss) the try catch with empty exception and review the unused variable. For the rest is awesome!

Thank you!

detouched commented:
Thanks for the quick reply :)

Sorry again that I wasn't looking at the up-to-date source code while making the change, my bad.
I'll drop the try-catch and the corresponding test, as well as the unused variable and update this PR.

detouched committed:
PR review - Fix and remove leftover

- No reload exceptions handling in TimedReloadStrategy
  (ConfigProvider is responsible for that)
- Remove unused variable
- Minor code format fix
